File: util/S3enc.java

import java.io.UnsupportedEncodingException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import sun.misc.BASE64Encoder;

public class S3enc {
	private static String generatePolicyDocumentString(String expiration, 
	String bucket, 
	String keyStartsWith, 
	String acl, 
	String successActionRedirect, 
	String contentLengthRangeMax) {
		return "{'expiration': '" + expiration + "'," +
              "'conditions': [ " +
                "{'bucket': '" + bucket + "'}, " +
                "['starts-with', '$key', '" + keyStartsWith + "']," +
                "{'acl': '" + acl + "'}," +
                "['content-length-range', 0, " + contentLengthRangeMax + "]" +
              "]" +
            "}";
	}

    public static void main(String[] args) throws UnsupportedEncodingException, NoSuchAlgorithmException, InvalidKeyException {
    	if (args.length!=7) {
    		System.out.println("usage:\n" + generatePolicyDocumentString("args[0]","args[1]","args[2]","args[3]","args[4]","args[5]"));
    		System.out.println("\nsecret key: args[6]");
    		return;
    	}
        String policy_document = generatePolicyDocumentString(args[0],args[1],args[2],args[3],args[4],args[5]);
        String aws_secret_key = args[6];
        
        String policy = (new BASE64Encoder()).encode(
                policy_document.getBytes("UTF-8")).replaceAll("\n", "").replaceAll("\r", "");

        Mac hmac = Mac.getInstance("HmacSHA1");
        hmac.init(new SecretKeySpec(
                aws_secret_key.getBytes("UTF-8"), "HmacSHA1"));
        String signature = (new BASE64Encoder()).encode(
                hmac.doFinal(policy.getBytes("UTF-8"))).replaceAll("\n", "");
        System.out.println("policy:" + policy);
        System.out.println("signature:" + signature);
    }
}

File: util/s3enc.py
#!/usr/bin/env python
# file: s3enc.py
import base64
import hmac, sha
import sys

def generatepolicydocument(expiration, bucket, keyprefix, acl, contentlengthrangemax):
	return """{{'expiration': '{0}',
	'conditions': [ 
		{{'bucket': '{1}'}}, 
		['starts-with', '$key', '{2}'],
		{{'acl': '{3}'}},
		['content-length-range', 0, {4}]
	]
}}""".format(expiration, bucket, keyprefix, acl, contentlengthrangemax)

def generatejson(accesskeyid, acl, policy, signature, thumbnailpath):
	return """
if (typeof djiaak === 'undefined') djiaak = {{}};
djiaak.settings = {{
	s3 : {{
		'AWSAccessKeyId': '{0}',
		'acl': '{1}',
		'policy': '{2}',
		'signature': '{3}'
	}},
	thumbnailPath: '{4}'
}};
""".format(accesskeyid, acl, policy, signature, thumbnailpath)

if len(sys.argv) != 8:
	sys.exit("invalid arguments. usage: s3enc.py [expiration] [bucket] [thumbnail path prefix] [acl] [content max size] [aws access key id] [aws secret key]")

policy = base64.b64encode(generatepolicydocument(sys.argv[1],sys.argv[2],sys.argv[3],sys.argv[4],sys.argv[5]))

signature = base64.b64encode(
    hmac.new(sys.argv[7], policy, sha).digest())
    
print(generatejson(sys.argv[6], sys.argv[4], policy, signature, sys.argv[3]))

